smart-connector/src/main/java/eu/knowledge/engine/smartconnector/impl/ReasonerProcessor.java

```diff
@@ -569,4 +601,163 @@ public CompletableFuture<Void> handle(eu.knowledge.engine.reasoner.api.BindingSe
+					LOG.info("Neighbor has antecedents, so check if the neighbor has gaps");
+
+					// knowledge engine specific code. We ignore meta knowledge interactions when
+					// looking for knowledge gaps, because they are very generic and make finding
+					// knowledge gaps nearly impossible.
+					boolean isMeta = isMetaKI(neighbor);
+
+					//TODO what if the graph contains loops?
+					//if (!isMeta && (someGaps = getKnowledgeGaps(neighbor)).isEmpty()) {
+					if (!isMeta && (someGaps1 = getKnowledgeGaps(neighbor)).isEmpty()) {
+						// found neighbor without knowledge gaps for the current triple, so current triple is covered.
+						LOG.info("Neighbor has no gaps");
+						foundNeighborWithoutGap = true;
+						break;
+					}
+					LOG.info("Neighbor has someGaps {}", someGaps);
+					collectedOrGaps.addAll(someGaps);
+					//new
+					collectedOrGaps1.addAll(someGaps1);
+				} else
+					foundNeighborWithoutGap = true;
+			}
+			LOG.info("Found a neighbor without gaps is {}", foundNeighborWithoutGap);
+
+			if (!foundNeighborWithoutGap) {
+				// there is a gap here, either in the current node or in a neighbor.
+
+				if (collectedOrGaps.isEmpty()) {
+					collectedOrGaps.add(new HashSet<>(Arrays.asList(entry.getKey())));
+				}
+				LOG.info("CollectedOrGaps is {}", collectedOrGaps);
+
+				if (collectedOrGaps1.isEmpty()) {
+					KnowledgeGap kg = new KnowledgeGap();
+					kg.add(entry.getKey());
+					collectedOrGaps1.add(kg);
+				}
+				LOG.info("CollectedOrGaps1 is {}", collectedOrGaps1);
+
+				Set<Set<TriplePattern>> newExistingOrGaps = new HashSet<>();
+				if (existingOrGaps.isEmpty()) {
+					existingOrGaps.addAll(collectedOrGaps);
+					LOG.info("Added collectedOrGaps to existingOrGaps");
+				} else {
+					Set<TriplePattern> newGap;
+					for (Set<TriplePattern> existingOrGap : existingOrGaps) {
+						for (Set<TriplePattern> collectedOrGap : collectedOrGaps) {
+							newGap = new HashSet<>();
+							newGap.addAll(existingOrGap);
+							newGap.addAll(collectedOrGap);
+							LOG.info("Found newGap {}", newGap);
+							newExistingOrGaps.add(newGap);
+						}
+					}
+					existingOrGaps = newExistingOrGaps;
+				}
+
+				//new
+				KnowledgeGapSet newExistingOrGaps1 = new KnowledgeGapSet();
+				if (existingOrGaps1.isEmpty()) {
+					existingOrGaps1.addAll(collectedOrGaps1);
+					LOG.info("Added collectedOrGaps1 to existingOrGaps1");
+				} else {
+					KnowledgeGap newGap1;
+					for (KnowledgeGap existingOrGap1 : existingOrGaps1) {
+						for (KnowledgeGap collectedOrGap1 : collectedOrGaps1) {
+							newGap1 = new KnowledgeGap();
+							newGap1.addAll(existingOrGap1);
+							newGap1.addAll(collectedOrGap1);
+							LOG.info("Found newGap1 {}", newGap1);
+							newExistingOrGaps1.add(newGap1);
+						}
+					}
+					existingOrGaps1 = newExistingOrGaps1;
+				}
+				// end new
+				
+			}
+		}
+		LOG.info("Found existingOrGaps {}", existingOrGaps);
+		LOG.info("Found existingOrGaps1 {}", existingOrGaps1);
```

Comment:
Consider moving most of these `info` level log statements to `debug` level.

Reply:
Done!